neotune: factoring improvements, progress notification

mlyle opened the pull request:
Pretty close to complete.

mlyle committed:
gcs: autotune: new beginning step class

Factor some validity logic into it, etc.

mlyle committed:
gcs: autotune: improve factoring a little

mlyle committed:
gcs: autotune: smush processing into wiz page

mlyle committed:
gcs: autotune: progress notification, etc

mlyle commented:
Not quite ready, need to send a couple signals.  Woops.

mlyle committed:
gcs: autotune: handle control of step advancement

mlyle commented:
OK, fixed up

mlyle committed:
gcs: autotune: add charts

mlyle commented:
![image](https://user-images.githubusercontent.com/90903/27452158-81fb1730-5746-11e7-9052-120cf4b96d37.png)

mlyle committed:
autotune: move to .ui form, prepare for progress

mlyle commented:
rebased

mlyle commented:
jenkins, test this please

tracernz commented:
Plot looks good. :+1:

tracernz commented:
Plot looks good. :+1:

jenkins, test this please

tracernz commented:
No ox again.

tracernz commented:
jenkins, build artifacts please

tracernz commented:
@mlyle there you go. 🙃

pug398 commented:
@mlyle I got this...
![neotune_warn](https://user-images.githubusercontent.com/3682100/27510014-abc6bcf4-58d6-11e7-86d0-34a62f559b19.PNG)

mlyle commented:
#1840 made WIndows GCS not work-- reverted.

jenkins, build artifacts please

mlyle commented:
oh crap, back to having conflicts, so i don't even want those set of artifacts.. re-rebasing

mlyle commented:
jenkins, build artifacts please

mlyle committed:
gcs: autotune: don't grab high tau false results

Only thing I've seen this affect is drawing of the yaw response graph in
GCS.

mlyle commented:
jenkins, build artifacts please

pug398 commented:
@mlyle Getting there. It only wipes out the config gadget now :)
![neo_warn](https://user-images.githubusercontent.com/3682100/27512395-5f31beda-590c-11e7-93ee-e430e835e77f.PNG)

amstan commented:
I ran an autotune on my quadcopter inside. Here's the autotune partition: [jun24.autotune.bin.zip](https://github.com/d-ronin/dRonin/files/1100040/jun24.autotune.bin.zip)

pug398 commented:
@amstan Are you on windows and get above warning message? The path looks good so I am unsure if it might be my machine.

mlyle commented:
@pug398 I think it's the effects of auto-updates on the windows builder.  Actual code is fine, it seems.

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
Still doesn't work.  I'm fairly stumped.  Here's what we know:

1. Windows ran updates on its own on the builder a couple days ago.  Yuck.
2. Before the revert earlier today, a huge set of modules didn't build.  With the revert, it is only this PR that doesn't build/only config plugin.
3. I tried rm -rf of the .ccache directory to see if it helped.  It doesn't.
4. I tried including charts at toplevel because it looks like the windeployqt stuff looks for it.  No dice.

tracernz commented on ground/gcs/src/app/app.pro:
```diff
@@ -6,6 +6,8 @@ TARGET = $$GCS_APP_TARGET
 DESTDIR = $$GCS_APP_PATH
 QT += xml
 QT += widgets
+QT += charts
```
~~What's this for?~~ :confused:
Oh, I see your comment above. This is definitely not needed. These are the modules drgcs.exe uses, not what the plugins use (they are completely independent shared libraries).

tracernz commented:
Need to put charts in here I think https://github.com/d-ronin/dRonin/blob/next/package/Makefile.win#L14

tracernz commented:
So, running in build environment works, but packages don't, or? And next packages are ok but not this branch?

tracernz commented on ground/gcs/src/plugins/config/config.pro:
```diff
@@ -5,6 +5,7 @@ DEFINES += QWT_DLL
 
 QT += svg
 QT += network
+QT += charts
```
This is correct.

tracernz committed:
Package:Win: Hax to deploy Qt5Charts.dll

windeployqt doesn't allow us to manually specify this plugin (obviously never bothered to support it properly)

tracernz commented:
jenkins, build artifacts please

tracernz commented:
jenkins, build artifacts please

tracernz commented:
@mlyle Those artifacts work fine. :wink: Just need to remove `Qt += charts` from app.pro.

tracernz commented:
Patch required to fix the issue: https://github.com/tracernz/qttools/commit/da9cb922532723891416214501d99ffeb21ff37a
I should probably attempt to push it upstream for 5.9.[2?].

mlyle commented:
Rebased out the QT += charts

jenkins, build artifacts please

mlyle commented:
No, I dind't, I screwed up the force push and didn't notice.  Now I did.  So much fail :(

Jenkins, build artifacts please

pug398 commented:
@mlyle no fails it's all good :)

mlyle commented:
finally.

@pug398 -- thanks for your patience.  :)

tracernz commented on ground/gcs/src/plugins/config/configautotunewidget.cpp:
```diff
@@ -395,417 +333,162 @@ void ConfigAutotuneWidget::stuffShareForm(AutotuneFinalPage *autotuneShareForm)
-        y2 = y1;
-        y1 = y;
+    // Used to track state across the entire wizard path.
+    AutotunedValues av;
```
Shadows the member variable? I would strongly prefer the member variable to have a better name since it has a _long_ lifetime and large scope, so probably best to change it to `tuneValues` or something.

mlyle replied:
I don't think it has a member variable of type AutotunedValues?

tracernz replied:
https://github.com/d-ronin/dRonin/pull/1835/files#diff-803cd9c265c5c58af660006e345bf45eR105

tracernz replied:
Ah, that's the page. Anyway, av is a bad name for that member variable IMO. Short meaningless names like that should be reserved for very tight scope like a loop var.

tracernz commented on ground/gcs/src/plugins/config/configautotunewidget.cpp:
```diff
@@ -395,417 +333,162 @@ void ConfigAutotuneWidget::stuffShareForm(AutotuneFinalPage *autotuneShareForm)
+            StabilizationSettings::GetInstance(getObjectManager());
+        Q_ASSERT(stabilizationSettings);
+        if (!stabilizationSettings)
+            return;
```
Can you add a qWarning here? That way we will know if there's a problem rather than it happening silently.

mlyle replied:
OK

tracernz commented on ground/gcs/src/plugins/config/configautotunewidget.cpp:
```diff
@@ -1036,3 +732,357 @@ AutotuneFinalPage::AutotuneFinalPage(QWidget *parent)
+    progressBar->setValue(0);
+
+    QMetaObject::invokeMethod(this, "doDownloadAndProcess", Qt::QueuedConnection);
+
```
Superfluous empty line.

mlyle replied:
OK

tracernz commented on ground/gcs/src/plugins/config/configautotunewidget.h:
```diff
@@ -27,28 +27,36 @@
 {
     bool valid;
 
+    QByteArray *data;
```
Reference preferred over pointer IMO.

mlyle replied:
If it's a reference, it needs to be known at the moment of struct construction, doesn't it?  I don't think it's easy to structure the code that way.

mlyle replied:
(I go back and forth on whether downloadFile should pass back a pointer to a constructed byte array or fill in something the user provides.  There's yuckiness each way)

tracernz replied:
No, the default initialiser will be used, which is an empty container (sizeof = sizeof(void *)). Assignment operator https://github.com/qt/qtbase/blob/d1ff9b419da5e3ace68479b02582222486ec9703/src/corelib/tools/qbytearray.h#L177 which simply assigns the d pointer from the other container (i.e. it costs the same as assigning a pointer). This is the way Qt containers are designed to be used.

tracernz replied:
Pardon me, should just be a QByteArray, not a pointer or ref.

mlyle replied:
OK, that makes more sense-- will do

tracernz commented on ground/gcs/src/plugins/uavtalk/telemetry.h:
```diff
@@ -84,7 +84,9 @@ class Telemetry : public QObject
     TelemetryStats getStats();
-    QByteArray *downloadFile(quint32 fileId, quint32 maxSize);
+    QByteArray *downloadFile(quint32 fileId, quint32 maxSize,
+            std::function<void(quint32)>progressCb = nullptr);
```
This is a bit ugly. Telemetry really needs to be in a different thread. :/

mlyle replied:
I don't know if I agree, but if we had a fully-async version it can all be signals (whether same thread or different).

it's just when it's synchronous, it seems dumb to connect/disconnect.

tracernz replied:
Not really, it would be consistent with the rest of the codebase whereas this is not.

mlyle replied:
I think QEventLoop is newer than a lot of the codebase, as are lambdas.  We also do relatively few things in the main UI  that are multistep-oriented... uploader, calibration, log download... and the wizards.  I think these would all be better written as eventloops (though cal should really be modal)

tracernz commented on ground/gcs/src/plugins/uavtalk/telemetry.cpp:
```diff
@@ -398,7 +398,12 @@ void Telemetry::transactionTimeout(ObjectTransactionInfo *transInfo)
-QByteArray *Telemetry::downloadFile(quint32 fileId, quint32 maxSize)
+/* This is synchronous, so we use a primitive callback mechanism
+ * instead of signal/slot.  Can have a future async variant if
+ * necessary
```
Necessary to shift to async in future IMO.

mlyle replied:
Probably for some things.

I really like intuitive code describing the sequence things will happen in instead of ... signals cascading to each other willy nilly and efforts to jam a crude state machine onto them.  So even if this becomes async, I think we'd want one level up to be synchronous/fairly imperative.

But conversely, we can use a lot of stack if there is any real concurrency between things using this model.

mlyle committed:
configautotunewidget: some changes from review

mlyle committed:
configautotunewidget: fairly extensive changes from feedback

mlyle committed:
gcs: autotune: average roll and pitch tau for now

mlyle committed:
gcs: autotune: do yaw & outer Ki by default

mlyle commented:
OK, this was a fair bit of work, but fixed up per comments.

Also made two minor functional changes:

1. For autotune, use the average of roll and pitch tau for now, instead of just roll tau.
2. Turn the long-standing experimental parameters that we tell everyone to turn on-- Yaw & Outer Ki-- on by default

jhitesma commented:
> Turn the long-standing experimental parameters that we tell everyone to turn on-- Yaw & Outer Ki-- on by default

👍   should help cut down on a good bit of confusion/concern from end users

mlyle commented:
Rebased to fix conflict.
